### PR TITLE
Avoid losing transitive packages in PrerequisiteProducer.computeDepedencies.

### PR DESCRIPTION
**Pull request by benjaminp**
The computeDependencies method always started with an empty transitive package set. If resolveConfiguredTargetDependencies had already been called and cached in the skyframe state on a previous restart, the transitive package set would not be properly filled. Resolve this by replacing the "transitive packages batch" with TransitiveDependencyState's incremental updateTransitivePackages method.

This bug appears to have been introduced by https://github.com/bazelbuild/bazel/commit/bd9eb4f3b505ca81622984e87df1c44829e04555.

### CODE REVIEW

**Comment by fmeum** [comment]
CC @greatfilter @aoeui

**Comment by greatfilter** [comment]
1. Can you add a failing test case written at a relatively high level of abstraction? This doesn't have much test coverage because `storeTransititivePackages` is not used in Google (this was a relatively recent change). I'm planning to delete `resolveConfiguredTargetDependencies` so any fix here would soon be deleted. The code that replaces it will only use `TransitiveDependencyState.updateTransitivePackages(ConfiguredTargetKey, NestedSet<Package>)`. 

2. Instead of putting the extra field in `PrerequisiteProducer.State` can you put it in `TransitiveDependencyState.PackageCollector`? It has better logical cohesion there. Note that this code would also be deleted when `resolveConfiguredTargetDependencies` is deleted.

**Comment by benjaminp** [comment]
I ruminated on the two valid issues you raise when I was preparing this change. Let me elaborate on my thinking: I came to make this CL after bisecting a strange new failure in a large monorepo to https://github.com/bazelbuild/bazel/commit/bd9eb4f3b505ca81622984e87df1c44829e04555. Perusing the diff, I noticed this bug and confirmed that fixing it resolved my original issue.

I could likely boil my problem down to an integration test. There is any irony in writing a test at a "high level of abstraction" for this issue, though, because any change in analysis/load skyframe code could silently invalidate the purpose of the test, adding one more useless, bloated, slow integration test to Bazel's pile.

I noticed the trend of consolidating transitive package collection into `PackageCollector`. In fact, I was wondering why bd9eb4f3b505ca81622984e87df1c44829e04555 added `PackageCollector.setTransitivePackages` rather than using `updateTransitivePackages` in CTF. I decided to stick with a minimal change, since I wasn't certain about the rationale of that design.

Basically, I selfishly assume that you will soon be rewriting this code, and this mediocre patch is functional enough in the mean time. Deleting `resolveConfiguredTargetDependencies` could completely obliterate this issue.

**Comment by greatfilter** [comment]
>I could likely boil my problem down to an integration test. There is any irony in writing a test at a "high level of abstraction" for this issue, though, because any change in analysis/load skyframe code could silently invalidate the purpose of the test, adding one more useless, bloated, slow integration test to Bazel's pile.

There's no coverage of incremental correctness of `storeTransitivePackages`. Lacking this coverage could cause problems in the future.

>I noticed the trend of consolidating transitive package collection into PackageCollector. In fact, I was wondering why https://github.com/bazelbuild/bazel/commit/bd9eb4f3b505ca81622984e87df1c44829e04555 added PackageCollector.setTransitivePackages rather than using updateTransitivePackages in CTF. I decided to stick with a minimal change, since I wasn't certain about the rationale of that design.

`SkyKeyComputeState` breaks generational hypothesis and increases the work done by the garbage collector. Adding a field there could be expensive than it appears because there are targets with 2 million `PrerequisiteProducer.State` instances. Putting it in `PackageCollector` makes it so the extra field doesn't add a tax to non-`storeTransitivePackages` builds. Using `updateTransitivePackages`, would cause repetitive `TreeMap.put`s on restarts which seemed to be more expensive than partially constructing NestedSets and throwing them away. Maybe that would have been simpler and not have this failure mode.

**Comment by benjaminp** [comment]
Very well; please see the latest changes.

**Comment by benjaminp** [comment]
https://github.com/bazelbuild/bazel/commit/fa28cf28f6a9210793039bbecef05018499b3b96
